packages/react-reconciler/src/ReactFiberCompleteWork.js

```diff
@@ -811,6 +817,44 @@ function completeWork(
 
         if (eventComponentInstance === null) {
           let responderState = null;
+          if (__DEV__ && !responder.allowMultipleHostChildren) {
+            let hostChildrenCount = 0;
+            const findAllHostChildren = node => {
+              if (node.tag === SuspenseComponent) {
+                const fallbackNode = isFiberSuspenseAndTimedOut(node)
+                  ? ((((node.child: any): Fiber).sibling: any): Fiber).child
+                  : ((node.child: any): Fiber).child;
+                if (fallbackNode === null) {
+                  return;
+                }
+                node = fallbackNode;
+              }
+              if (
+                node.tag === HostComponent ||
+                node.tag === HostText ||
+                node.tag === HostPortal
+              ) {
+                hostChildrenCount++;
+              } else {
+                const child = node.child;
+                if (child !== null) {
+                  findAllHostChildren(child);
+                }
+              }
+              const sibling = node.sibling;
+              if (sibling !== null) {
+                findAllHostChildren(sibling);
+              }
+            };
+            if (workInProgress.child !== null) {
+              findAllHostChildren(workInProgress.child);
+            }
+            warning(
+              hostChildrenCount < 2,
+              'An event component "%s" was found to have multiple host children.',
```

Comment:
Little detail. I think the message implies that all event components must have a single child. Could be phrased something like…

```suggestion
              'A "<%s>" event component cannot contain multiple host children.',
```

I think the UX consequences could be severe enough to cause a SEV if ignored. Should we throw an error like `React.Children.only` does?

Reply:
An error means that we have to move this out of the DEV condition. I wanted to avoid this as this check is expensive.

Reply:
It can't be a DEV-only error?

Reply:
This is too expensive to do in prod IMO.

Reply:
I guess I can move back to the older version that didn't use recursion and make it an invariant, it just feels like a relatively expensive check to do on every fiber complete phase – especially as the tree below is likely to be quite large before the first host component is known.